### PR TITLE
StringValues Add IsNull, always set single item to _value

### PR DESCRIPTION
**Pull request by benaadams**

### CODE REVIEW

**Comment by davidfowl** [comment]
@benaadams did you have to reformat all of the methods 🤣

**Comment by benaadams** [comment]
>  did you have to reformat all of the methods 🤣

Better? 😄

**Comment by benaadams** [comment]
Then can use `.IsNull` in Kestrel rather than `.Count == 0` for less branchy fasterness

**Comment by davidfowl** [comment]
Any perf results?

**Comment by benaadams** [comment]
> Any perf results?

This is from your point in https://github.com/aspnet/KestrelHttpServer/pull/2347#discussion_r171105414

So the perf would be from a follow on change in Kestrel changing header existence checks from the equivalent of 
```csharp
if ((_value != null ? 1 : (_values?.Length ?? 0)) > 0)
```
to
```csharp
if (!(_value == null && _values == null))
```

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 1]
```diff
@@ -27,8 +27,23 @@ public StringValues(string value)
 
         public StringValues(string[] values)
         {
-            _value = null;
-            _values = values;
+            if (values == null || values.Length == 0)
```
Comment (halter73) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: Maybe add a comment explaining that if _values isn't set to null when values.Length is zero, the IsNull property will be broken.

Comment (benaadams) [reply src/Microsoft.Extensions.Primitives/StringValues.cs]: Added

**Comment by benaadams** [comment]
CI was slow...

**Comment by benaadams** [comment]
Added change to remove nulls from arrays to cover the corner case of nulls getting in via `string[]`

This also means the `if (value != null)` check per item per header on output in Kestrel could be dropped

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 2]
```diff
@@ -27,8 +27,71 @@ public StringValues(string value)
...
+            }
+            else if (values.Length == 1)
+            {
+                // Set single item arrays to the string value
+                _values = null;
+                _value = values[0];
+            }
+            else
+            {
+                // Remove any nulls from the array else counts will be off
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: This is beyond StringValues' purview to modify the input like this. And as far as we know this is a rare corner case so why put the extra effort into it? It's not even worth writing the tests for. None of your other changes in this PR depend on this filter, correct?

Comment (benaadams) [reply src/Microsoft.Extensions.Primitives/StringValues.cs]: Its an edge case with consequences.

The `string[]` ctor is the only way `null` can appear with a count
```csharp
new StringValues(new []{(string)null, (string)null, (string)null })
```
Will report `.Count == 3`

However, it will also output zero headers; so is it really 3 headers?

As `null` can get into `StringValues` values every header value has to be null checked in Kestrel to see if it is null or not before outputting it.

This means the previous will report `.Count == 0` and all the null checks can be removed from Kestrel for header output.

Comment (benaadams) [reply src/Microsoft.Extensions.Primitives/StringValues.cs]: Common case is there aren't nulls and the `string[]` ctor is not used; so its pushing to cost for potential null values coming from that .ctor into the .ctor; rather than having guards spread everywhere.

Comment (Tratcher) [reply src/Microsoft.Extensions.Primitives/StringValues.cs]: The constructor is called from an implicit converter, it shouldn't be doing non-trivial transformations. StringValues also only wraps the array so you can't prevent those values from becoming null after the constructor.

Since Kestrel needs the values to not be null then it should do a one time filter, possibly via a helper method provided by StringValues like WithoutNulls or GetNonNulls.

Comment (benaadams) [reply src/Microsoft.Extensions.Primitives/StringValues.cs]: > Since Kestrel needs the values to not be null then it should do a one time filter, possibly via a helper method provided by StringValues like WithoutNulls or GetNonNulls.

That wouldn't achieve any real benefit over what exists now (and likely would be more expensive)

Reverted

**Comment by halter73** [comment]
Still LGTM

File: src/Microsoft.Extensions.Primitives/StringValues.cs [hunk 3]
```diff
@@ -420,12 +421,12 @@ public override bool Equals(object obj)
         {
             if (obj == null)
             {
-                return Equals(this, StringValues.Empty);
+                return this.IsNull;
             }
 
-            if (obj is string)
+            if (obj is string st)
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringValues.cs]: nit: str

Comment (benaadams) [reply src/Microsoft.Extensions.Primitives/StringValues.cs]: fixed

**Comment by halter73** [comment]
Always set single item to _value

**Comment by halter73** [comment]
Whoops. I didn't squash this. Sorry about the git history.

**Comment by Tratcher** [comment]
FYI: This change regressed StringValues.Empty. It used to be equivalent to new string[0], but now it's null. It causes the same problem for MVC's ValueProviderResult.None. @rynowak 
https://github.com/aspnet/Mvc/blob/1aeaf69f87cbf91f4f56183eb1254f69a50b406a/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs#L35

**Comment by Tratcher** [comment]
@benaadams heads up. We're going to revert this and reintroduce it more cautiously by doing full universe builds before merging.

**Comment by benaadams** [comment]
> We're going to revert this and reintroduce it more cautiously by doing full universe builds before merging.

Makes sense, a broken build is no place to be...

**Comment by Tratcher** [comment]
Reverted. https://github.com/aspnet/Common/commit/43ad60b85c04f91bdcc08016891d382a87a3117f
